File: src/server/import.rs
use std::path::Path;

use crate::server::{db, markdown, mistral, qdrant, state::AppState};

/// Run the import pipeline on all `.md` files in `data/`.
pub async fn run_import(state: &AppState) -> Result<(), Box<dyn std::error::Error>> {
    let data_dir = Path::new("data");
    let processed_dir = data_dir.join("processed");
    let failed_dir = data_dir.join("failed");

    tokio::fs::create_dir_all(&processed_dir).await?;
    tokio::fs::create_dir_all(&failed_dir).await?;

    // Ensure Qdrant collection exists
    qdrant::ensure_collection(&state.qdrant).await?;

    let mut entries = tokio::fs::read_dir(data_dir).await?;
    while let Some(entry) = entries.next_entry().await? {
        let path = entry.path();
        if path.extension().and_then(|e| e.to_str()) != Some("md") {
            continue;
        }

        let filename = path
            .file_name()
            .unwrap_or_default()
            .to_string_lossy()
            .to_string();

        tracing::info!("processing {filename}");

        match import_file(state, &path).await {
            Ok(()) => {
                let dest = processed_dir.join(&filename);
                tokio::fs::rename(&path, &dest).await?;
                tracing::info!("{filename} -> processed/");
            }
            Err(e) => {
                tracing::error!("failed to import {filename}: {e}");
                let dest = failed_dir.join(&filename);
                tokio::fs::rename(&path, &dest).await?;
            }
        }
    }

    Ok(())
}

async fn import_file(state: &AppState, path: &Path) -> Result<(), Box<dyn std::error::Error>> {
    let raw = tokio::fs::read_to_string(path).await?;
    let parsed = markdown::parse_markdown(&raw).map_err(|e| e.to_string())?;
    let fm = &parsed.frontmatter;

    // Check if book already exists
    let existing = db::find_book_by_ref_id(&state.pool, &fm.id).await?;

    match existing {
        Some((_id, ref existing_hash)) if existing_hash == &parsed.hash => {
            tracing::info!("book {} unchanged, skipping", fm.id);
            return Ok(());
        }
        _ => {}
    }

    // Generate summary via Mistral if not in frontmatter
    let summary = if fm.summary.is_some() {
        fm.summary.clone()
    } else if !state.mistral_api_key.is_empty() {
        let s = mistral::generate_summary(
            &state.http_client,
            &state.mistral_api_key,
            &parsed.content[..parsed.content.len().min(6000)],
        )
        .await
        .ok();
        s
    } else {
        None
    };

    // Upsert book in PostgreSQL
    let book_id = match existing {
        Some(_) => {
            db::update_book(
                &state.pool,
                &fm.id,
                &parsed.hash,
                &fm.title,
                &fm.authors,
                fm.editor.as_deref(),
                &fm.tags,
                fm.edition_date.as_deref(),
                summary.as_deref(),
                fm.introduction.as_deref(),
                fm.cover_text.as_deref(),
                fm.ean.as_deref(),
                fm.isbn.as_deref(),
                &parsed.content,
                &fm.reseller_paper_urls,
                &fm.reseller_digital_urls,
            )
            .await?
        }
        None => {
            db::insert_book(
                &state.pool,
                &fm.id,
                &parsed.hash,
                &fm.title,
                &fm.authors,
                fm.editor.as_deref(),
                &fm.tags,
                fm.edition_date.as_deref(),
                summary.as_deref(),
                fm.introduction.as_deref(),
                fm.cover_text.as_deref(),
                fm.ean.as_deref(),
                fm.isbn.as_deref(),
                &parsed.content,
                &fm.reseller_paper_urls,
                &fm.reseller_digital_urls,
            )
            .await?
        }
    };

    // Extract chapters and generate chapter summaries
    let chapters = markdown::extract_chapters(&parsed.content);

    if !state.mistral_api_key.is_empty() {
        let chapter_inputs: Vec<(Option<String>, String)> = chapters
            .iter()
            .map(|c| (c.title.clone(), c.text.clone()))
            .collect();

        let chapter_summaries = mistral::generate_chapter_summaries(
            &state.http_client,
            &state.mistral_api_key,
            &chapter_inputs,
        )
        .await
        .unwrap_or_default();

        let summaries_data: Vec<(i32, Option<&str>, &str)> = chapter_summaries
            .iter()
            .enumerate()
            .filter(|(_, s)| !s.is_empty())
            .map(|(i, s)| {
                let title = chapters.get(i).and_then(|c| c.title.as_deref());
                (i as i32, title, s.as_str())
            })
            .collect();

        db::upsert_chapter_summaries(&state.pool, book_id, &summaries_data).await?;
    }

    // Chunk content and generate embeddings
    let chunks = markdown::chunk_text(&chapters);

    if !state.mistral_api_key.is_empty() && !chunks.is_empty() {
        let texts: Vec<String> = chunks.iter().map(|c| c.text.clone()).collect();

        let embeddings = mistral::embed_texts(
            &state.http_client,
            &state.mistral_api_key,
            &texts,
        )
        .await
        .map_err(|e| format!("embedding error: {e}"))?;

        // Delete old points if updating
        if existing.is_some() {
            qdrant::delete_book_points(&state.qdrant, book_id).await?;
        }

        let chunk_data: Vec<(usize, Option<String>, usize, String)> = chunks
            .iter()
            .map(|c| {
                (
                    c.chapter_idx,
                    c.chapter_title.clone(),
                    c.chunk_index,
                    c.text.clone(),
                )
            })
            .collect();

        qdrant::upsert_chunks(
            &state.qdrant,
            book_id,
            &fm.id,
            &fm.title,
            &fm.authors,
            &fm.tags,
            &chunk_data,
            &embeddings,
        )
        .await?;
    }

    tracing::info!("imported book {} ({})", fm.id, fm.title);
    Ok(())
}

File: scripts/import.ts
import fs from "fs/promises";
import path from "path";
import {
  findBookByReference,
  insertBook,
  updateBook,
  upsertChapterSummaries,
} from "../src/lib/db";
import { parseMarkdown, extractChapters, chunkText } from "../src/lib/markdown";
import {
  embedTexts,
  generateSummary,
  generateChapterSummaries,
} from "../src/lib/mistral";
import {
  ensureCollection,
  deleteBookPoints,
  upsertChunks,
} from "../src/lib/qdrant";

const DATA_DIR = path.resolve("data");
const PROCESSED_DIR = path.join(DATA_DIR, "processed");
const FAILED_DIR = path.join(DATA_DIR, "failed");

async function main() {
  await fs.mkdir(PROCESSED_DIR, { recursive: true });
  await fs.mkdir(FAILED_DIR, { recursive: true });

  await ensureCollection();

  const entries = await fs.readdir(DATA_DIR);
  const mdFiles = entries.filter((f) => f.endsWith(".md"));

  for (const filename of mdFiles) {
    const filePath = path.join(DATA_DIR, filename);
    console.log(`processing ${filename}`);

    try {
      await importFile(filePath);
      const dest = path.join(PROCESSED_DIR, filename);
      await fs.rename(filePath, dest);
      console.log(`${filename} -> processed/`);
    } catch (e) {
      console.error(`failed to import ${filename}:`, e);
      const dest = path.join(FAILED_DIR, filename);
      await fs.rename(filePath, dest);
    }
  }
}

async function importFile(filePath: string) {
  const raw = await fs.readFile(filePath, "utf-8");
  const parsed = parseMarkdown(raw);
  const fm = parsed.frontmatter;

  // Check if book already exists
  const existing = await findBookByReference(fm.reference);

  if (existing && existing.hash === parsed.hash) {
    console.log(`book ${fm.reference} unchanged, skipping`);
    return;
  }

  // Generate summary via Mistral if not in frontmatter
  const mistralKey = process.env.MISTRAL_API_KEY ?? "";
  let summary = fm.summary;

  if (!summary && mistralKey) {
    try {
      summary = await generateSummary(parsed.content.slice(0, 6000));
    } catch (e) {
      console.error("summary generation failed:", e);
    }
  }

  // Upsert book in PostgreSQL
  let bookId: string;
  if (existing) {
    bookId = await updateBook(
      fm.reference,
      parsed.hash,
      fm.title,
      fm.authors,
      fm.editor,
      fm.tags,
      fm.edition_date,
      summary ?? null,
      fm.introduction,
      fm.cover_text,
      fm.ean,
      fm.isbn,
      parsed.content,
      fm.reseller_paper_urls,
      fm.reseller_digital_urls,
    );
  } else {
    bookId = await insertBook(
      fm.reference,
      parsed.hash,
      fm.title,
      fm.authors,
      fm.editor,
      fm.tags,
      fm.edition_date,
      summary ?? null,
      fm.introduction,
      fm.cover_text,
      fm.ean,
      fm.isbn,
      parsed.content,
      fm.reseller_paper_urls,
      fm.reseller_digital_urls,
    );
  }

  // Extract chapters and generate chapter summaries
  const chapters = extractChapters(parsed.content);

  if (mistralKey) {
    const chapterInputs = chapters.map((c) => ({
      title: c.title,
      text: c.text,
    }));

    const chapterSummaries = await generateChapterSummaries(
      chapterInputs,
    ).catch(() => [] as string[]);

    const summariesData = chapterSummaries
      .map((s, i) => ({
        chapterIdx: i,
        title: chapters[i]?.title ?? null,
        summary: s,
      }))
      .filter((s) => s.summary);

    await upsertChapterSummaries(bookId, summariesData);
  }

  // Chunk content and generate embeddings
  const chunks = chunkText(chapters);

  if (mistralKey && chunks.length > 0) {
    const texts = chunks.map((c) => c.text);
    const embeddings = await embedTexts(texts);

    // Delete old points if updating
    if (existing) {
      await deleteBookPoints(bookId);
    }

    await upsertChunks(
      bookId,
      fm.reference,
      fm.title,
      fm.authors,
      fm.tags,
      chunks,
      embeddings,
    );
  }

  console.log(`imported book ${fm.reference} (${fm.title})`);
}

main().catch((e) => {
  console.error("import failed:", e);
  process.exit(1);
});
